std/experimental/checkedint.d

```diff
@@ -0,0 +1,1584 @@
+        }
+    }
+
+    // opBinaryRight
+    /**
+    */
+    auto opBinaryRight(string op, Lhs)(const Lhs lhs)
+    if (isIntegral!Lhs || isFloatingPoint!Lhs || is(Lhs == bool))
+    {
+        static if (hasMember!(Hook, "hookOpBinaryRight"))
+        {
+            auto r = hook.hookOpBinaryRight!op(lhs, payload);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else static if (hasMember!(Hook, "hookOpBinary"))
+        {
+            auto r = hook.hookOpBinary!op(lhs, payload);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else static if (is(Lhs == bool))
+        {
+            return mixin("ubyte(lhs)"~op~"this");
+        }
+        else static if (isFloatingPoint!Lhs)
+        {
+            return mixin("lhs"~op~"payload");
+        }
+        else static if (hasMember!(Hook, "onOverflow"))
+        {
+            bool overflow;
+            auto r = opChecked!op(lhs, T(payload), overflow);
+            if (overflow) r = hook.onOverflow!op(42);
+            return Checked!(typeof(r), Hook)(r);
+        }
+        else
+        {
+            // Default is built-in behavior
+            auto r = mixin("lhs"~op~"T(payload)");
+            return Checked!(typeof(r), Hook)(r);
+        }
+    }
+
+    // opOpAssign
+    /**
+    */
+    ref Checked opOpAssign(string op, Rhs)(const Rhs rhs)
+    if (isIntegral!Rhs || isFloatingPoint!Rhs || is(Rhs == bool))
+    {
+        static assert(is(typeof(mixin("payload"~op~"=rhs")) == T));
+
+        static if (hasMember!(Hook, "hookOpOpAssign"))
+        {
+            hook.hookOpOpAssign!op(payload, rhs);
+        }
+        else
+        {
+            alias R = typeof(payload + rhs);
+            auto r = mixin("this"~op~"rhs").payload;
+
+            static if (valueConvertible!(R, T) ||
+                !hasMember!(Hook, "onBadOpOpAssign") ||
+                op.among(">>", ">>>"))
+            {
+                // No need to check these
+                payload = cast(T) r;
+            }
+            else
+            {
+                static if (isUnsigned!T && !isUnsigned!R)
+                    // Example: ushort += int
+                    const bad = r < 0 || r > max.payload;
+                else
+                    // Some narrowing is afoot
+                    static if (R.min < min.payload)
+                        // Example: int += long
+                        const bad = r > max.payload || r < min.payload;
+                    else
+                        // Example: uint += ulong
+                        const bad = r > max.payload;
+                if (bad)
+                    payload = hook.onBadOpOpAssign!op(payload, Rhs(rhs));
+                else
+                    payload = cast(T) r;
+            }
+        }
+        return this;
+    }
+}
+
+// representation
+unittest
+{
+    assert(Checked!(ubyte, void)(ubyte(22)).representation == 22);
+}
+
+/**
+Force all overflows to fail with `assert(0)`.
+*/
+struct Croak
```

Comment:
Needs a better name. Croak is not very descriptive.

Maybe something like `OverflowError` or `ForceAssert`.